File: backend/app/schemas/certificate_validity.py
"""
Certificate Validity Schemas
Part 14 — Certificate Validity Monitoring for BidVerify AI
Pydantic response and request models for certificate monitoring.
"""

from datetime import date, datetime
from typing import Any, Dict, List, Optional
import uuid
from pydantic import BaseModel, ConfigDict, Field


class DocumentValidityDTO(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    document_id: uuid.UUID
    bid_id: Optional[uuid.UUID] = None
    organization_id: uuid.UUID
    document_name: Optional[str] = None
    document_type: str
    issue_date: Optional[date] = None
    expiry_date: Optional[date] = None
    validity_status: str
    days_until_expiry: Optional[int] = None
    date_source: str
    source_page: Optional[int] = None
    source_text: Optional[str] = None
    confidence: float
    is_current: bool
    submission_validity_status: Optional[str] = None
    last_checked_at: datetime
    next_check_at: Optional[datetime] = None
    metadata_json: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class CertificateValidityStatsDTO(BaseModel):
    total_monitored: int = 0
    valid_count: int = 0
    expiring_soon_count: int = 0
    expired_count: int = 0
    no_expiry_count: int = 0
    review_required_count: int = 0
    unknown_count: int = 0


class BidderCertificateListResponse(BaseModel):
    items: List[DocumentValidityDTO]
    total: int
    page: int
    page_size: int
    total_pages: int
    stats: CertificateValidityStatsDTO


class ProcurementCertificateListResponse(BaseModel):
    items: List[DocumentValidityDTO]
    total: int
    page: int
    page_size: int
    total_pages: int


class CertificateValidityRecheckResponse(BaseModel):
    record: DocumentValidityDTO
    message: str


class PeriodicValidityCheckResponse(BaseModel):
    total_checked: int
    status_transitions: int
    status_breakdown: Dict[str, int]
    reference_date: str

File: frontend/src/types/certificate_validity.ts
/**
 * Certificate Validity Monitoring Types
 * Part 14 — Certificate Validity Monitoring for BidVerify AI
 */

export enum ValidityStatus {
  VALID = "VALID",
  EXPIRING_SOON = "EXPIRING_SOON",
  EXPIRED = "EXPIRED",
  NO_EXPIRY = "NO_EXPIRY",
  UNKNOWN = "UNKNOWN",
  REVIEW_REQUIRED = "REVIEW_REQUIRED",
}

export enum ValidityDateSource {
  STRUCTURED_EXTRACTION = "STRUCTURED_EXTRACTION",
  VERIFICATION_ADAPTER = "VERIFICATION_ADAPTER",
  MANUAL_OVERRIDE = "MANUAL_OVERRIDE",
}

export interface DocumentValidityItem {
  id: string;
  document_id: string;
  bid_id?: string | null;
  organization_id: string;
  document_name?: string | null;
  document_type: string;
  issue_date?: string | null;
  expiry_date?: string | null;
  validity_status: ValidityStatus | string;
  days_until_expiry?: number | null;
  date_source: string;
  source_page?: number | null;
  source_text?: string | null;
  confidence: number;
  is_current: boolean;
  submission_validity_status?: string | null;
  last_checked_at: string;
  next_check_at?: string | null;
  metadata_json: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface CertificateValidityStats {
  total_monitored: number;
  valid_count: number;
  expiring_soon_count: number;
  expired_count: number;
  no_expiry_count: number;
  review_required_count: number;
  unknown_count: number;
}

export interface BidderCertificateListResponse {
  items: DocumentValidityItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  stats: CertificateValidityStats;
}

export interface ProcurementCertificateListResponse {
  items: DocumentValidityItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface CertificateValidityRecheckResponse {
  record: DocumentValidityItem;
  message: string;
}

export interface PeriodicValidityCheckResponse {
  total_checked: number;
  status_transitions: number;
  status_breakdown: Record<string, number>;
  reference_date: string;
}
